### PR TITLE
Worklist #21391 Implement a Parenting Tool

### PR DESCRIPTION
**Pull request by jmo7**
Commit consists of two javascript files, createParentator.js and parentator.js, along with necessary media files supplied by HighFidelity. A test plan in pdf format is also attached.
[TestPlan_for_21391.pdf](https://github.com/highfidelity/hifi/files/1128989/TestPlan_for_21391.pdf)

### CODE REVIEW

**Comment by CainFoool** [comment]
Testing shortly.

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1396/HighFidelity-Beta-PR10902-0f44e67c88ce9b308f74be71141b4c344d706d9c.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1396/HighFidelity-Beta-PR10902-0f44e67c88ce9b308f74be71141b4c344d706d9c.dmg

**Comment by CainFoool** [comment]
Testing.

**Comment by YuppyPlays** [comment]
Testing.

**Comment by YuppyPlays** [comment]
Run into a small issue:
quoting from the PDF:
Tests
1) We will now check that the parent-ator appears properly
1.1) Choose Edit -> Open and Run Script File (or Ctrl + O).
1.2) Navigate to <HIGH_FIDELITY_ROOT>/unpublishedScripts/parent-ator and choose the
createParentator.js:
a) Check that the parent-ator entity appears in front of you (purple gun-like thing).

**I cannot seem to see a unpublishedscripts folder? or the script itself anywhere**

**Comment by cozza13** [comment]
@YuppyPlays use this script https://rawgit.com/jmo7/hifi/0f44e67c88ce9b308f74be71141b4c344d706d9c/unpublishedScripts/parent-ator/createParentator.js

**Comment by YuppyPlays** [comment]
Testing.

**Comment by YuppyPlays** [comment]
Testing complete:
Sections from the PDF:
1) **Pass**
2) **Pass**
3) **Pass**
4) **Pass**
5) **Pass**
6) **Fail? Not sure if I didn't get a correct domain, but in Dev-Welcome and in Voxeltopia, It first brought up the message upon spawning "Looks like you don't have permissions to edit in this domain" but after tapping a few objects some will allow me to parent them, some locked, etc. Does not give the message regarding not having permissions except for when first rezing the tool.**

**Comment by Nex-Pro** [comment]
I see a number of issues with this. 

I'm getting really bad sound effects with it. There should be 4 I only hear some really distorted noise. 
Cleared cache but did not fix.

The 'Parentator' model rezzes pointing at the user's face. Would probably be better to add 180 to START_POSITION.y then normalize.

The 'Parentator' model floats away.  Would suggest increasing linear damping from the default. Maybe go to 0.9 or more.

**Comment by Nex-Pro** [comment]
Also  if the user taps an object   then releases and re-equips with the parentator   it continues to display "tap the child"  but will set the object tapped first as the child of the first other object tapped after re-equipping.  So the messages change from Tap the child to the "Yay" success message.

**Comment by Nex-Pro** [comment]
I noticed that a parent relationship got broken using this. Will try to get a repro.

**Comment by Nex-Pro** [comment]
The face normals are inverted on the three rings surrounding the barrel of the Parentator model.

**Comment by Nex-Pro** [comment]
Parenting can be broken with this. On occasions the Parentator seems to get out of sequence.
With a group of objects eventually one or more can get orphaned. In effect (occasionally ) a single first tap can produce the success message.

**Comment by Nex-Pro** [comment]
There is an error in the description on line 7 of parentator.js

**Comment by Nex-Pro** [comment]
Equipping can fail with a smaller than normal avatar as the offsets are hard coded rather than scaled relative the the user's scale.

**Comment by Nex-Pro** [comment]
@jmo7 Overall this is a good and useful tool. Welcome aboard.

**Comment by kencooke** [comment]
@jmo7 The reason your sounds are noisy/distorted is because the WAVs are 24-bit, but we require 16-bit. Your WAV headers look like:
```
format tag = 0xfffe
sample rate = 48000
channels = 2
bits = 24
```

You also want to use mono WAVs for spatialized sounds. Stereo is intended for non-diegetic sounds (music, etc.)

File: unpublishedScripts/parent-ator/createParentator.js [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+var scriptURL = Script.resolvePath('parentator.js');
+var MODEL_URL = Script.resolvePath('resources/Parent-Tool-Production.fbx');
+var COLLISION_HULL_URL = Script.resolvePath('resources/Parent-Tool-CollisionHull.obj');
+//var COLLISION_SOUND_URL = 'http://hifi-production.s3.amazonaws.com/DomainContent/Toybox/ping_pong_gun/plastic_impact.L.wav';
+var START_POSITION = Vec3.sum(Vec3.sum(MyAvatar.position, {
+  x: 0,
+  y: 0.5,
+  z: 0
+}), Vec3.multiply(0.7, Quat.getForward(Camera.getOrientation())));
+var START_ROTATION = Vec3.sum(MyAvatar.position, Vec3.multiply(1.5, Quat.getFront(Camera.getOrientation())));
```
Comment (AndrewMeadows) [inline unpublishedScripts/parent-ator/createParentator.js]: The START_ROTATION doesn't look right... it is a Vec3.sum() which means it is not a Quat (quaternion).  Maybe you just want something like:
`var START_ROTATION = MyAvatar.orientation;`

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1985/HighFidelity-Beta-PR10902-635be3e2e98410e194bfe85f9993a7715e4e5a72.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1985/HighFidelity-Beta-PR10902-635be3e2e98410e194bfe85f9993a7715e4e5a72.dmg

**Comment by jmo7** [comment]
I've updated this branch (21391) with the fixes for the issues mentioned.

**Comment by themelissabrown** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1988/HighFidelity-Beta-PR10902-635be3e2e98410e194bfe85f9993a7715e4e5a72.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1988/HighFidelity-Beta-PR10902-635be3e2e98410e194bfe85f9993a7715e4e5a72.exe

**Comment by Nex-Pro** [comment]
Looking at this again

**Comment by Nex-Pro** [comment]
### Dynamic Moving objects that bump into the parentator can become linked. 
**( TESTED AND CONFIRMED IN DESKTOP AND HMD)**

To repro: Throw a couple of dynamic objects at the parentator either in desktop or HMD.
Observe that they become linked.

Test notes were found to be incorrect and this is noted below.

### Test Plan for #21391 Implement a Parenting Tool


This is written so that both newbies as well as pros could do this testing. At various steps additional

instructions are provided for those newer to HighFidelity and are in blue, like the fiollowing:
Steps to check an object’s (entity’s) properties


1) Open your tablet.

2) Click the “Create” button.

3) In the VR world (not on your tablet) select the object whose properties you want to check.
4) Go back to the tablet and click the “Properties” tab on top of the screen.
5) Scroll to find the appropriate property.

Preparation Steps

1) In HighFidelity create 8 boxes and or spheres.
Set them so that:
1 box/sphere is locked (a.k.a “locked 1”)
3 boxes/spheres are dynamic (a.k.a. “dynamic 1”, “dynamic 2”, and “dynamic 3”)
4 boxes/spheres are not dynamic or locked (a.k.a. “regular 1”, “regular 2”, “regular 3”, and “regular 4”)

1) In HighFidelity open your tablet and choose the "Create" button. Choose “Box” or “Sphere”. Move your avatar’s
perspective somewhat (for instance turn around somewhat). If you stay still and create a new object it will be created in
the exact same space as your previous object).
2) Repeat step 1 seven more times.
3) Making sure that one of the boxes/spheres you just created is selected, on your tablet choose the Properties tab at
the top.
4) Check the "locked" checkbox in the list of Properties (it’s one of the first ones).
5) Select a different box or sphere in the vr world, and make sure your tablet is on the properties screen. Scroll down on
the tablet to the “dynamic” property and set the checkbox on.
6) Repeat step 5 for two more boxes or spheres.
----------------------------------------------------------------------------

### Tests


1) We will now check that the parent-ator appears properly

1.1) Choose Edit -> Open and Run Script File (or Ctrl + O).

1.2) Navigate to <HIGH_FIDELITY_ROOT>/unpublishedScripts/parent-ator and choose the
createParentator.js: Used Raw link as script not present in build. 
https://raw.githubusercontent.com/jmo7/hifi/635be3e2e98410e194bfe85f9993a7715e4e5a72/unpublishedScripts/parent-ator/createParentator.js

a) Check that the parent-ator entity appears in front of you (purple gun-like thing). **PASS**

b) Check that you can equip the parent-ator to both your left **PASS** and and right **PASS** hand and make sure
 its  positioning in your hand looks ok. **Reasonable with Mathew avatar**

c) Check that as you equip the parent-ator a sound is played. **PASS**

d) Check that the parent-ator’s screen says "Tap the child object". **PASS**

----------------------------------------------------------------------------

2) We will now check that it can parent to items together

2.1) Tap the end of the parent-ator on “regular 1”.

a) Check that the parent-ator screen now says "Now tap the parent object".**PASS**

b) Check that a sound is played.**PASS**


2.2) Tap the parent-ator on “regular 2”.

a) Check that the parent-ator screen now says "Yay! Your two objects are now parented".**PASS**

b) Check that a sound is played.**PASS**

c) Check that after 5 seconds the parent-ator screen returns to the initial text of "Tap the child
 object".**PASS** Also plays a sound( should probably have been noted in this testing guide )


2.3) Select “regular 1”.

a) Check that the parentID property is set to the id of “regular 2”. **PASS**

b) Grab and move “regular 1” and see that “regular 2” moves also. ( ***IMO*** ( more than two years using HiFi) ***This instruction is incorrect*** In this case regular 2 becomes the parent and moving regular 1 should not move regular 2. Moving regular 2 should move regular 1)***. Pass on corrected instructions. **DOUBLE CHECK THIS** 

----------------------------------------------------------------------------

3) We will now check that the parent-ator can re-parent items appropriately

3.1) Tap parent-ator on “regular 1”.

a) Check that the parent-ator screen now says "Now tap the parent object". **PASS**

b) Check that a sound is played.**PASS**


3.2) Tap the parent-ator on “regular 3”.

a) Check that the parent-ator screen now says "Yay! Your two objects are now parented".**PASS**

b) Check that a sound is played.**PASS**

c) Check that after 5 seconds the parent-ator screen returns to the initial text of “Tap the child object". **PASS**

d) Check that the parentID of “regular 1” is now set to the id of “regular 3” (instead of the old
 parentID of “regular 2”). **PASS**

----------------------------------------------------------------------------

4) We will now make sure that a dynamic item that is tapped first (set as the child) becomes nondynamic (and the second item tapped (the parent) stays at whatever it was -- dynamic or non-dynamic)

4.1) Tap the parent-ator on “dynamic 1”.

a) Check that the object’s dynamic property has been unchecked.**PASS** Had to repeat this step as parentator cannot be used in 'Create' mode


4.2) Tap the parent-ator on “dynamic 2”.

a) Check that the object’s dynamic property is still checked as dynamic. **PASS**


4.3) Tap the parent-ator on a “dynamic 3”.

a) Check that the object’s dynamic property has been unchecked. **PASS**


4.2) Tap the parent-ator on “regular 4”.

a) Check that the object’s dynamic property is still unchecked. **PASS**

----------------------------------------------------------------------------

5) We will now make sure that a locked object can not be chosen

5.1) Tap the parent-ator on the “locked 1”.

a) Check that the parent-ator’s screen says “That item is locked – try another”. **PASS**  Also plays a sound( should probably have been noted in this testing guide )


5.2 Tap the parent-ator on “regular 1” and then tap “locked 1” as the parent.

a) Check that the parent-ator’s screen says “That item is locked – try another” . **PASS**

----------------------------------------------------------------------------

6) We will now make sure that if you are in a domain where you don’t have permissions to create
objects (“rez entities”), you are alerted to this and nothing further happens

6.1) Open your tablet, and choose “Goto”.

6.2) Pick a domain on the list other than one of the Workshop domains (formerly known as
“Welcome” domains). We’re trying to find a domain where you don’t have rights to edit items.

6.3) After the domain has loaded, run the parent-ator script.

6.4) Equip your avatar with the parent-ator

6.5) Tap the parent-ator on an object in the domain.
a) Check that the parent-ator’s screen says “Looks like you don’t have permission to edit in this
 domain...”. **PASS**



### Observations/Feedback

Had to use the raw link to the script.

Sounds are now fixed. But my friend and I found them to be annoying

Parentator Tool pretty much works as expected.

The sounds are great. ONCE. The sounds are annoying when this is used for multiple operations especially to other users.

**However the tool remains active even when not being held.**  This can have consequences. 

The sounds are annoying in continous use.

The gun could potentially bump into two objects and parent the first to the second. 

Moving objects that bump into the parentator could become linked. ( TESTED AND CONFIRMED IN DESKTOP AND HMD)

Did I mention that my friend and I found the sounds annoying ?

The delay between touching child touching parent and the 'available to use' sound ( I guess that's what it is ) is way too long. 

Oh and the sounds .... they are really, really annoying.

Apart from that Well done.

Oh and the sounds....

**Comment by Nex-Pro** [comment]
@jmo7 ^

**Comment by Nex-Pro** [comment]
@jmo7 @themelissabrown @kencooke Can't the sounds be made local to the user ? This is like the finger particles/gestures thing in that surely only the user needs that feedback.

**Comment by Nex-Pro** [comment]
I feel the need to fly through an area full of entities while holding this thing ! Should be fun.

**Comment by kencooke** [comment]
Keeping the sounds local to the user is probably a good idea. To do this, set localOnly to true:

```
var audioProperties = { 
    volume: 0.3, 
    position: Entities.getEntityProperties(this.entityID).position, 
    localOnly: true
}
```

**Comment by themelissabrown** [comment]
@jmo7 one change this needs before we make it live is to make sure it only functions when equipped. Once it is unequipped it should be disabled.
We should also check that an item touched by it is not/does not remain equipped. Let me know if you have questions.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2219/HighFidelity-Beta-PR10902-dfb60273475cdce0d5c15f9f31d1eee8d98f1e98.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2219/HighFidelity-Beta-PR10902-dfb60273475cdce0d5c15f9f31d1eee8d98f1e98.exe

File: unpublishedScripts/parent-ator/createParentator.js [hunk 2]
```diff
@@ -9,17 +9,15 @@
...
-var START_POSITION = Vec3.sum(Vec3.sum(MyAvatar.position, {
-  x: 0,
-  y: 0.5,
-  z: 0
-}), Vec3.multiply(0.7, Quat.getForward(Camera.getOrientation())));
-var START_ROTATION = Vec3.sum(MyAvatar.position, Vec3.multiply(1.5, Quat.getFront(Camera.getOrientation())));
+
+// the fbx model needs to be rotated from where it would naturally face when it first initializes
+var ROT_Y_180 = {x: 0, y: 1, z: 0, w: 0};
+var START_ROTATION = Quat.normalize(Quat.multiply(ROT_Y_180, Camera.getOrientation()));
```
Comment (AndrewMeadows) [inline unpublishedScripts/parent-ator/createParentator.js]: The short version: you need to swap the order of the `Quat.multiply()` arguments here.

The long version:

In general rotations don't commute:  `A*B != B*A`.  However they do when both rotations are about the same axis... which is usually the case in High Fidelity right now... which is why your code here "works".

Unfortunately for you now, but good news that we are finally getting around to it, we are starting to track down those places in the code where `up = yAxis` is assumed.  We have to generalize such code to handle arbitrary `up` in order to support avatars walking around on spherical planets or standing on arbitrarily rotated floors.  The code as you've written it here doesn't exactly "assume" the `up = yAxis` condition, but works accidentally when true.

Either you look at the code and say "Oh, he's right!" or you say "Why does he think so?".  For the latter case I offer this explanation:  The **180 flip about Y** needs to be applied in the local-frame of the object.  Rotations operate on vectors **from the left** and you want the **180 flip about Y** to apply **first** therefore it must be **rightmost**.

File: unpublishedScripts/parent-ator/createParentator.js [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+//  See the accompanying file LICENSE or http://www.apache.org/licenses/LICENSE-2.0.html
+//
+
+var scriptURL = Script.resolvePath('parentator.js');
+var MODEL_URL = Script.resolvePath('resources/Parent-Tool-Production.fbx');
+var COLLISION_HULL_URL = Script.resolvePath('resources/Parent-Tool-CollisionHull.obj');
+
+// the fbx model needs to be rotated from where it would naturally face when it first initializes
+var ROT_Y_180 = {x: 0, y: 1, z: 0, w: 0};
+var START_ROTATION = Quat.normalize(Quat.multiply(ROT_Y_180, Camera.getOrientation()));
```
Comment (AndrewMeadows) [inline unpublishedScripts/parent-ator/createParentator.js]: The short version: you need to swap the order of the `Quat.multiply()` arguments here.

The long version:

In general rotations don't commute: `A*B != B*A`. However they do when both rotations are about the same axis... which is usually the case in High Fidelity right now... which is why your code here "works".

Unfortunately for you now, but good news that we are finally getting around to it, we are starting to track down those places in the code where `up = yAxis` is assumed. We have to generalize such code to handle arbitrary up in order to support avatars walking around on spherical planets or standing on arbitrarily rotated floors. The code as you've written it here doesn't exactly "assume" the `up = yAxis` condition, but only works accidentally when true.

Either you look at the code and say "Oh, he's right!" or you say "Why does he think so?". For the latter case I offer this explanation: The **180 flip about Y** needs to be applied in the **local-frame** of the object. Rotations operate on vectors **from the left** and you want the **180 flip about Y** to apply **first** therefore it must be **rightmost**.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2317/HighFidelity-Beta-PR10902-41c8c5da9978b4435be365bf2d257ae11c792821.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2317/HighFidelity-Beta-PR10902-41c8c5da9978b4435be365bf2d257ae11c792821.exe

**Pull request merged**